src/libraries/Microsoft.PowerFx.Core/Public/Config/DeferredSymbolTable.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Types.Enums;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Provides symbol table with deferred semantics. 
+    /// This is semantically readonly - there is a fixes set of known symbols, just the type information is populated lazily.
+    /// Good when we can quickly load the names (to populate the intellisense completion list), but then load remaining symbol details on-demand. 
+    /// </summary>
+    [DebuggerDisplay("{DebugName}")]
+    internal class DeferredSymbolTable : ReadOnlySymbolTable, IGlobalSymbolNameResolver, INameResolver
```

Comment:
This is the key new class. 
It's internal - it's created via static helpers on ReadOnlySymbolTable, like the other derived symbols.